pkg/install/upgrade_test.go

```diff
@@ -20,80 +20,135 @@ import (
-		},
-		{
-			name:    "higher version, no update needed",
-			fakecli: newFakecli("", "4.4.5"),
-		},
-		{
-			name:           "lower version, update needed (4.4)",
-			fakecli:        newFakecli("", "4.4.1"),
+			name: "upgrade to Y latest",
+			fakecli: newFakecli(configv1.ClusterVersionStatus{
+				Desired: configv1.Update{
+					Version: "4.3.1",
+				},
+				Conditions: []configv1.ClusterOperatorStatusCondition{
+					{
+						Type:   configv1.OperatorAvailable,
+						Status: configv1.ConditionTrue,
+					},
+				},
+			}),
 			wantUpdated:    true,
-			desiredVersion: Stream44.Version.String(),
+			desiredVersion: stream43.Version.String(),
 		},
 		{
-			name:    "higher version, no update needed",
-			fakecli: newFakecli("", "4.3.19"),
+			name: "no upgrade, Y higher than expected",
+			fakecli: newFakecli(configv1.ClusterVersionStatus{
+				Desired: configv1.Update{
+					Version: "4.3.99",
+				},
+				Conditions: []configv1.ClusterOperatorStatusCondition{
+					{
+						Type:   configv1.OperatorAvailable,
+						Status: configv1.ConditionTrue,
+					},
+				},
+			}),
+			wantUpdated: false,
+			wantErr:     "not upgrading: cvo desired version is 4.3.99",
 		},
 		{
-			name:           "lower version, update needed (3.3)",
-			fakecli:        newFakecli("", "4.3.14"),
-			wantUpdated:    true,
-			desiredVersion: Stream43.Version.String(),
+			name: "no upgrade, Y match but unhealthy cluster",
+			fakecli: newFakecli(configv1.ClusterVersionStatus{
+				Desired: configv1.Update{
+					Version: stream43.Version.String(),
+				},
+				Conditions: []configv1.ClusterOperatorStatusCondition{
+					{
+						Type:   configv1.OperatorAvailable,
+						Status: configv1.ConditionFalse,
+					},
+				},
+			}),
+			wantUpdated: false,
+			wantErr:     "not upgrading: previous upgrade in-progress",
 		},
 		{
-			name:           "on a channel, update needed",
-			fakecli:        newFakecli("my-channel", "4.3.14"),
-			desiredVersion: Stream43.Version.String(),
+			name: "upgrade, Y match",
+			fakecli: newFakecli(configv1.ClusterVersionStatus{
+				Desired: configv1.Update{
+					Version: stream43.Version.String(),
+				},
+				Conditions: []configv1.ClusterOperatorStatusCondition{
+					{
+						Type:   configv1.OperatorAvailable,
+						Status: configv1.ConditionTrue,
+					},
+				},
+			}),
 			wantUpdated:    true,
+			desiredVersion: stream44.Version.String(),
 		},
 		{
-			name:           "on a channel, no update needed",
-			fakecli:        newFakecli("my-channel", "4.4.4"),
-			desiredVersion: Stream44.Version.String(),
+			name: "upgrade, Y match 2",
+			fakecli: newFakecli(configv1.ClusterVersionStatus{
+				Desired: configv1.Update{
+					Version: stream44.Version.String(),
+				},
+				Conditions: []configv1.ClusterOperatorStatusCondition{
+					{
+						Type:   configv1.OperatorAvailable,
+						Status: configv1.ConditionTrue,
+					},
+				},
+			}),
+			wantUpdated:    true,
+			desiredVersion: stream45.Version.String(),
```

Comment:
Might want to add a comment that this behaviour isn't currently available in production (as stream 4.5 doesn't exist).

Reply:
Not sure if this is needed. It just tests future behaviour. It would be same as If I would be testing with 42,43,44. It just proves logic is sane